server/src/main/java/io/druid/query/lookup/LookupReferencesManager.java

```diff
@@ -417,81 +467,37 @@ private void loadAllLookupsAndInitStateRef()
   {
     ImmutableMap.Builder<String, LookupExtractorFactoryContainer> builder = ImmutableMap.builder();
     ExecutorService executorService = Execs.multiThreaded(
         lookupConfig.getNumLookupLoadingThreads(),
         "LookupReferencesManager-Startup-%s"
     );
-    ExecutorCompletionService completionService = new ExecutorCompletionService(executorService);
-    List<Future<Map.Entry>> futures = new ArrayList<>();
+    CompletionService<Map.Entry<LookupBean, LookupExtractorFactoryContainer>> completionService =
+        new ExecutorCompletionService<>(executorService);
+    List<Future<?>> allFutures = new ArrayList<>();
     try {
       LOG.info("Starting lookup loading process");
-      for (LookupBean lookupBean : lookupBeanList) {
-        futures.add(
-            completionService.submit(
-                () -> {
-                  LookupExtractorFactoryContainer container = lookupBean.getContainer();
-                  LOG.info(
-                      "Starting lookup [%s]:[%s]",
-                      lookupBean.getName(),
-                      container
-                  );
-                  if (container.getLookupExtractorFactory().start()) {
-                    LOG.info(
-                        "Started lookup [%s]:[%s]",
-                        lookupBean.getName(),
-                        container
-                    );
-                    return new AbstractMap.SimpleImmutableEntry<>(lookupBean.getName(), container);
-                  } else {
-                    LOG.error(
-                        "Failed to start lookup [%s]:[%s]",
-                        lookupBean.getName(),
-                        container
-                    );
-                    return null;
-                  }
-                }
-            )
-        );
-      }
-      for (int i = 0; i < futures.size(); i++) {
-        try {
-          final Future<AbstractMap.SimpleImmutableEntry<String, LookupExtractorFactoryContainer>> completedFuture = completionService
-              .take();
-          final AbstractMap.SimpleImmutableEntry<String, LookupExtractorFactoryContainer> lookupResult = completedFuture
-              .get();
-          if (lookupResult != null) {
-            builder.put(lookupResult);
-          }
-        }
-        catch (ExecutionException e) {
-          LOG.error(e, "Execution error during lookup loading.");
+      for (int i = 0; i < lookupConfig.getLookupStartRetries(); i++) {
+        LOG.info("Round of attempts #%d, [%d] lookups", i + 1, lookupBeanList.size());
+        Pair<Map<LookupBean, LookupExtractorFactoryContainer>, List<LookupBean>> results =
+            startLookups(lookupBeanList, completionService, allFutures);
+        Map<LookupBean, LookupExtractorFactoryContainer> successfulLookups = results.lhs;
+        successfulLookups.forEach((lookupBean, container) -> builder.put(lookupBean.getName(), container));
+        List<LookupBean> failedLookups = results.rhs;
```

Comment:
the lookup bean list can then simply be filtered down by excluding the ones whose names are that are in the "success" map.